org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/config/ConfigurationExtension.java

```diff
@@ -0,0 +1,156 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit5.config;
+
+import static org.osgi.test.common.inject.FieldInjector.findAnnotatedFields;
+import static org.osgi.test.common.inject.FieldInjector.setField;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.test.common.annotation.config.Configuration;
+import org.osgi.test.common.annotation.config.Configurations;
+import org.osgi.test.common.dictionary.Dictionaries;
+
+public class ConfigurationExtension implements BeforeEachCallback, ParameterResolver {
+
+	final static Namespace NAMESPACE = Namespace.create(ConfigurationExtension.class);
+
+	@Override
+	public void beforeEach(ExtensionContext extensionContext) throws Exception {
+
+		List<Field> fields = findAnnotatedFields(extensionContext.getRequiredTestClass(), Configuration.class);
+
+		BundleContext bundleContext = FrameworkUtil.getBundle(extensionContext.getRequiredTestClass())
+			.getBundleContext();
+
+		fields.forEach(field -> {
+			assertValidFieldCandidate(field);
+
+			Configuration configAnnotation = field.getAnnotation(Configuration.class);
+			Class<?> memberType = field.getType();
+			Type genericMemberType = field.getGenericType();
+
+			setField(field, extensionContext.getRequiredTestInstance(),
+				resolveReturnValue(configAnnotation, bundleContext, extensionContext));
+		});
+
+		extensionContext.getElement()
+			.ifPresent((element) -> {
+				Configurations configAnnotations = element.getAnnotation(Configurations.class);
+				if (configAnnotations != null) {
+					Stream.of(configAnnotations.value())
+						.forEachOrdered((configAnnotation) -> {
+							resolveReturnValue(configAnnotation, bundleContext, extensionContext);
+						});
+				}
+				Configuration configAnnotation = element.getAnnotation(Configuration.class);
+				if (configAnnotation != null) {
+					resolveReturnValue(configAnnotation, bundleContext, extensionContext);
+				}
+			});
+	}
```

Comment:
I've recently tidied up and factored out some of the equivalent code in `BundleContextExtension` - have a look and see if you can reuse it. Some of the code now lives here: https://github.com/osgi/osgi-test/blob/6a756341e12349faebd61a43acb14a38e7368cc8/org.osgi.test.common/src/main/java/org/osgi/test/common/inject/FieldInjector.java#L216-L236